### PR TITLE
pyloxi: Connection class

### PR DESCRIPTION
**Pull request by rlane**
Reviewer: @kenchiang

This class creates a thread which continually parses OpenFlow messages off the 
supplied socket and places them in a queue. The class has methods for reading
messages from the RX queue, sending messages, and higher level operations like
request-response and multipart transactions.

I'm putting this in Loxigen because it's so commonly used even though it goes 
beyond the usual LOXI scope of serialization.

### CODE REVIEW

**Comment by bsn-abat** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.bigswitch.com//job/loxigen_pull_req/140/
Test PASSed.

**Comment by rlane** [comment]
ping @kenchiang

File: py_gen/templates/connection.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+        offset = 0
+        while offset < len(buf):
+            if offset + 8 > len(buf):
+                # Not enough data for the OpenFlow header
+                break
+
+            # Parse the header to get type
+            hdr_version, hdr_type, hdr_length, hdr_xid = loxi.of14.message.parse_header(buf[offset:])
+
+            # Use loxi to resolve to ofp of matching version
```
Comment (kenchiang) [inline py_gen/templates/connection.py]: minor: 'use loxi to resolve ofp of matching version' ?

File: py_gen/templates/connection.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+            rawmsg = buf[offset : offset + hdr_length]
+            offset += hdr_length
+
+            msg = ofp.message.parse_message(rawmsg)
+            if not msg:
+                self.logger.warn("Could not parse message")
+                continue
+
+            self.logger.debug("Received message %s.%s xid %d length %d",
+                                type(msg).__module__, type(msg).__name__, hdr_xid, hdr_length)
```
Comment (kenchiang) [inline py_gen/templates/connection.py]: minor: indent

File: py_gen/templates/connection.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+            buf = recvd
+
+        offset = 0
+        while offset < len(buf):
+            if offset + 8 > len(buf):
+                # Not enough data for the OpenFlow header
+                break
+
+            # Parse the header to get type
+            hdr_version, hdr_type, hdr_length, hdr_xid = loxi.of14.message.parse_header(buf[offset:])
```
Comment (kenchiang) [inline py_gen/templates/connection.py]: hdr_length seems to be unfortunately named; it's the message length as specified in the header, but when quickly read seems to imply 'header length' instead.

File: py_gen/templates/connection.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+OpenFlow connection class
+
+This class creates a thread which continually parses OpenFlow messages off the
+supplied socket and places them in a queue. The class has methods for reading messages
+from the RX queue, sending messages, and higher level operations like request-response
+and multipart transactions.
+"""
+
+import loxi
+import loxi.of14
```
Comment (kenchiang) [inline py_gen/templates/connection.py]: will we need to keep bumping the version upwards as new openflow versions are released?

**Comment by kenchiang** [comment]
Overall, LGTM.

@wilmo119 @harshsin Can you python experts take a look?  Thanks.

**Comment by rlane** [comment]
Thanks for the review. Pushed a couple of commits.

File: py_gen/templates/connection.py (hunk 4 above)
Comment (rlane) [inline py_gen/templates/connection.py]: Probably not. We don't want to break callers relying on OF 1.4 features that aren't present in later versions.

**Comment by bsn-abat** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.bigswitch.com//job/loxigen_pull_req/148/
Test PASSed.

**Comment by kenchiang** [comment]
new commits look OK.  please feel free to poke @wilmo119 or @harshsin if you want a more thorough review.  thanks.

**Comment by rlane** [comment]
SELF-ACK

**Comment by bsn-abat** [comment]
ABAT: START_MERGE:  Log file at http://vss1/abat/2015.05.27.1808-m.loxigen.master/abat.log

**Comment by bsn-abat** [comment]
ABAT: ACCEPT:  Successfully merged

In case you want to see the build log, check out :
Log file at http://vss1/abat/2015.05.27.1808-m.loxigen.master/abat.log
/cc

**Pull request closed without merging**